### PR TITLE
Database: psycopg2 to psycopg migration #6669

### LINKED ISSUE #6669: `psycopg2-binary` should not be used in production - `psycopg2` should be used instead
### Description

For PostgreSQL-related functions, Rucio uses the `psycopg2-binary` library:

https://github.com/rucio/rucio/blob/339744866e9550f0f068f8e3c9574a95ccaa7d59/requirements.txt#L34

However, the [`psycopg2` docs](https://www.psycopg.org/docs/install.html#psycopg-vs-psycopg-binary) say:
> The psycopg2-binary package is meant for beginners to start playing with Python and PostgreSQL without the need to meet the build requirements.
> If you are the maintainer of a published package depending on psycopg2 you shouldn’t use psycopg2-binary as a module dependency. For production use you are advised to use the source distribution.

See also: https://stackoverflow.com/questions/70330567/what-is-the-different-about-psycopg2-and-psycopg2-binary-python-package

Additionally there is also [`psycopg3`](https://www.psycopg.org/psycopg3/), which introduces many new features (the differences between `psycopg2` and `psycopg3` are listed [here](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html)), like support for static typing.

To do:
- If possible, we might want to consider upgrading from `psycopg2` to `psycopg3`
- If not, at the very least we should move from `psycopg2-binary` to `psycopg2` as recommended in their docs.

This is currently a blocker for this issue: https://github.com/rucio/rucio/issues/6666, which requires the [`psycopg2.sql`](https://www.psycopg.org/docs/sql.html) module.

### Steps to reproduce

N/A

### Rucio Version

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by Geogouz**
This PR upgrades the PostgreSQL driver from psycopg2 to psycopg (3). There has already been some work to tackle this (see #6993) but since I am following a quite different approach and the integration of those commits would yield a messy branch, I am starting a new one here and we can continue the discussion here.

A few important remarks:

I have added functionality to handle and parse requirements with “extras” (e.g., psycopg[binary,pool]).

While all tests are currently passing, we should remain cautious. Since `Psycopg` serves as the underlying driver for all `PostgreSQL` connections (used both indirectly through `SQLAlchemy` and directly in some parts of the code), there might still be edge cases that our tests have not detected.

For instance, I had to explicitly do casts of the port value at endpoints to an integer because `Psycopg` enforces stricter type handling compared to `Psycopg2`. The following method:

https://github.com/rucio/rucio/blob/594959f33b7d91892bee4d2147c7bbe1ec8b4dde/lib/rucio/core/rse.py#L1584-L1591

even though expects an optional integer for port, our tests (and possibly even production usage) are passing a string without raising an error in other contexts.

Additionally, the `postgres_meta.py` module directly interacts with Psycopg (bypassing SQLAlchemy and its models). I had to update the SQL statements to align with Psycopg’s stricter requirements. These changes call maybe for more thorough testing to ensure that all functionality remains intact and behaves as expected for this plugin.

### CODE REVIEW

**Comment by rdimaio** [comment]
> even though expects an optional integer for port, our tests (and possibly even production usage) are passing a string without raising an error in other contexts.

that's due to the `@transactional_session` decorator, which accidentally hides the function signature from linters, see https://github.com/rucio/rucio/issues/6986 for more context

**Comment by robbarnsley** [comment]
Thanks @Geogouz. Apologies for taking so long to get around to this.

As far as I am aware, we (SKAO) are the only ones using the external postgres plugin & we're not running a production instance currently so the... "blast radius" resulting from any resulting issues around this will be quite limited.

That being said, I haven't seen any problems with my (admittedly cursory) checks of getting/setting metadata following these changes, and the the (limited) unit tests all pass, so, from that perspective i'm happy for this to be merged.

**Review by robbarnsley — DISMISSED** [review]
LGTM!

**Comment by bari12** [comment]
Please rebase and adapt commit messages to the style recommended in the https://rucio.cern.ch/documentation/contributing/

File: lib/rucio/core/did_meta_plugins/postgres_meta.py [hunk 1]
```diff
@@ -97,17 +101,18 @@ def _try_create_metadata_table(self):
...
-            ("name", "varchar", "NOT NULL"),
-            ("data", "jsonb", "DEFAULT", "'{}'::jsonb"),
-            ("UNIQUE", "(scope, name)")  # unique scope+name table constraint, required for ON CONFLICT
+        table_clauses = [
+            sql.SQL("id bigint NOT NULL GENERATED ALWAYS AS IDENTITY"),
+            sql.SQL("vo varchar NOT NULL"),
+            sql.SQL("scope varchar NOT NULL"),
+            sql.SQL("name varchar NOT NULL"),
+            sql.SQL("data jsonb DEFAULT '{}'::jsonb"),
+            sql.SQL("UNIQUE (scope, name)")
```
Comment (rdimaio) [inline lib/rucio/core/did_meta_plugins/postgres_meta.py]: I think we should keep the `# unique scope+name table constraint, required for ON CONFLICT` for clarity

Comment (Geogouz) [reply lib/rucio/core/did_meta_plugins/postgres_meta.py]: Okie. I'll adapt also the commit messages. Pushed it for the moment to see how tests are doing.

**Review by rdimaio — APPROVED** [review]

**Comment by bari12** [comment]
Merged as feature for 37
